File: NetFramework/DashboardLayout/VB/DashboardLayoutDemo/SampleHelper.vb
Imports System
Imports System.Linq
Imports System.IO
Imports System.Data
Imports System.Data.OleDb
Imports System.Windows.Forms
Imports C1.Win.C1Themes
Imports System.Runtime.InteropServices

Namespace DashboardDemo
    Public Class SampleHelper
        Implements IDisposable

        Private _data As DataSet = New DataSet()
        Private _layoutsDir As String = Directory.GetCurrentDirectory() & "\Layouts"
        Public Property Theme As String

        Public ReadOnly Property DefaultLayoutPath As String
            Get
                Dim di = New DirectoryInfo(Directory.GetCurrentDirectory() & "\Resources")
                If Not di.Exists Then di.Create()
                Return Directory.GetCurrentDirectory() & "\Resources\DefaultLayout.xml"
            End Get
        End Property

        Public Function GetSourceTable(ByVal tableName As String, ByVal selectCommandText As String) As DataTable
            Dim dt As DataTable = New DataTable(tableName)

            Using da = New OleDbDataAdapter(selectCommandText, GetConnectionString())
                da.Fill(dt)
                _data.Tables.Add(dt)
            End Using

            Return dt
        End Function

        Public Sub CenterPosition(ByVal owner As Control, ByVal child As Control)
            If owner IsNot Nothing AndAlso child IsNot Nothing Then
                Dim x As Integer = (owner.Width - child.Width) / 2
                Dim y As Integer = (owner.Height - child.Height) / 2
                child.Location = New System.Drawing.Point(If(x > 0, x, 0), If(y > 0, y, 0))
            End If
        End Sub

        Public Function SaveLayoutDialog(<Out> ByRef filePath As String) As Boolean
            filePath = Nothing

            Using dialog = New SaveLayoutForm()
                If Theme IsNot Nothing Then C1ThemeController.ApplyThemeToControlTree(dialog, C1ThemeController.GetThemeByName(Theme, False))

                If dialog.ShowDialog() = DialogResult.OK Then
                    If Not Directory.Exists(_layoutsDir) Then Directory.CreateDirectory(_layoutsDir)
                    filePath = _layoutsDir & "\" & dialog.FileName & ".xml"
                    Return True
                End If
            End Using

            Return False
        End Function

        Public Sub ManageLayoutDialog()
            Using dialog = New ManageLayoutForm(Me)
                If Theme IsNot Nothing Then C1ThemeController.ApplyThemeToControlTree(dialog, C1ThemeController.GetThemeByName(Theme, False))
                dialog.ShowDialog()
            End Using
        End Sub

        Public Function GetLayouts() As String()
            Dim di = New DirectoryInfo(_layoutsDir)

            If di.Exists Then
                Dim files = di.GetFiles("*.xml")
                Dim names = files.[Select](Function(x) x.Name)
                Dim array = names.ToArray()
                Return di.GetFiles("*.xml").[Select](Function(x) x.Name).ToArray()
            End If

            Return Nothing
        End Function

        Public Sub DeleteLayout(ByVal name As String)
            If Directory.Exists(_layoutsDir) Then File.Delete(GetLayoutPath(name))
        End Sub

        Public Sub RenameLayout(ByVal name As String)
            Using dialog = New SaveLayoutForm(name)
                dialog.Text = "Rename Layout"
                If Theme IsNot Nothing Then C1ThemeController.ApplyThemeToControlTree(dialog, C1ThemeController.GetThemeByName(Theme, False))
                If dialog.ShowDialog() = DialogResult.OK AndAlso Directory.Exists(_layoutsDir) Then File.Move(GetLayoutPath(name), GetLayoutPath(dialog.FileName))
            End Using
        End Sub

        Public Function GetLayoutPath(ByVal name As String) As String
            Return _layoutsDir & "\" & name
        End Function

        Private Function GetConnectionString() As String
            Dim path As String = Environment.GetFolderPath(Environment.SpecialFolder.Personal) & "\ComponentOne Samples\Common"
            Dim conn As String = "provider=microsoft.jet.oledb.4.0;data source={0}\c1nwind.mdb;"
            Return String.Format(conn, path)
        End Function

        Public Shared _instance As SampleHelper

        Public Shared ReadOnly Property Instance As SampleHelper
            Get
                Return If(_instance, (CSharpImpl.__Assign(_instance, New SampleHelper())))
            End Get
        End Property

        Public Shared Sub DisposeInstance()
            _instance.Dispose()
            _instance = Nothing
        End Sub

        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
            If disposing Then

                If _data IsNot Nothing Then

                    For Each table As System.IDisposable In _data.Tables
                        table.Dispose()
                    Next

                    _data.Dispose()
                End If
            End If
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub

        Private Class CSharpImpl
            <Obsolete("Please refactor calling code to use normal Visual Basic assignment")>
            Shared Function __Assign(Of T)(ByRef target As T, value As T) As T
                target = value
                Return value
            End Function
        End Class
    End Class
End Namespace

File: Next/DashboardLayout/CS/DashboardLayoutDemo/SampleHelper.cs
using System;
using System.Linq;
using System.IO;
using System.Data;
using System.Data.OleDb;
using System.Windows.Forms;
using C1.Win.Themes;

namespace DashboardDemo
{
    public class SampleHelper: IDisposable
    {
        #region Fields
        private DataSet _data = new DataSet();
        private string _layoutsDir = Directory.GetCurrentDirectory() + @"\Layouts";
        #endregion Fields

        #region Model
        public string Theme { get; set; }
        public string DefaultLayoutPath
        {
            get
            {
                var di = new DirectoryInfo(Directory.GetCurrentDirectory() + @"\Resources");
                if (!di.Exists)
                    di.Create();
                return Directory.GetCurrentDirectory() + @"\Resources\DefaultLayout.xml";
            }
        }
        public DataTable GetSourceTable(string tableName, string selectCommandText)
        {
            DataTable dt = new DataTable(tableName);
            using (var da = new OleDbDataAdapter(selectCommandText, GetConnectionString()))
            {
                da.Fill(dt);
                _data.Tables.Add(dt);
            }
            return dt;
        }
        public void CenterPosition(Control owner, Control child)
        {
            if (owner != null && child != null)
            {
                int x = (owner.Width - child.Width) / 2;
                int y = (owner.Height - child.Height) / 2;
                child.Location = new System.Drawing.Point(x > 0 ? x : 0, y > 0 ? y : 0);
            }
        }
        public bool SaveLayoutDialog(out string filePath)
        {
            filePath = null;
            using (var dialog = new SaveLayoutForm())
            {
                if(Theme != null)
                    C1ThemeController.ApplyThemeToControlTree(dialog, C1ThemeController.GetThemeByName(Theme, false));
                if (dialog.ShowDialog() == DialogResult.OK)
                {
                    if(!Directory.Exists(_layoutsDir))
                        Directory.CreateDirectory(_layoutsDir);
                    filePath = _layoutsDir + @"\" + dialog.FileName + ".xml";
                    return true;
                }
            }
            return false;
        }
        public void ManageLayoutDialog()
        {
            using (var dialog = new ManageLayoutForm(this))
            {
                if (Theme != null)
                    C1ThemeController.ApplyThemeToControlTree(dialog, C1ThemeController.GetThemeByName(Theme, false));
                dialog.ShowDialog();
            }
        }
        public string[] GetLayouts()
        {
            var di = new DirectoryInfo(_layoutsDir);
            if (di.Exists)
            {
                var files = di.GetFiles("*.xml");
                var names = files.Select(x => x.Name);
                var array = names.ToArray();
                return di.GetFiles("*.xml").Select(x => x.Name).ToArray();
            }
            return null;
        }
        public void DeleteLayout(string name)
        {
            if (Directory.Exists(_layoutsDir))
                File.Delete(GetLayoutPath(name));
        }
        public void RenameLayout(string name)
        {
            using (var dialog = new SaveLayoutForm(name))
            {
                dialog.Text = "Rename Layout";
                if (Theme != null)
                    C1ThemeController.ApplyThemeToControlTree(dialog, C1ThemeController.GetThemeByName(Theme, false));
                if (dialog.ShowDialog() == DialogResult.OK && Directory.Exists(_layoutsDir))
                    File.Move(GetLayoutPath(name), GetLayoutPath(dialog.FileName));
            }
        }
        public string GetLayoutPath(string name) => _layoutsDir + @"\" + name;
        #endregion Model

        #region Private
        private string GetConnectionString()
        {
            string path = Environment.GetFolderPath(Environment.SpecialFolder.Personal) + @"\ComponentOne Samples\Common";
            string conn = @"provider=microsoft.jet.oledb.4.0;data source={0}\c1nwind.mdb;";
            return string.Format(conn, path);
        }
        #endregion Private

        #region Singleton 
        public static SampleHelper _instance;

        public static SampleHelper Instance => _instance ?? (_instance = new SampleHelper());

        public static void DisposeInstance()
        {
            _instance.Dispose();
            _instance = null;
        }
        #endregion

        #region IDisposable
        /// <summary>
        /// Releases all unmanaged resources.
        /// </summary>
        /// <param name="disposing"></param>
        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
            {
                if (_data != null)
                {
                    foreach (System.IDisposable table in _data.Tables)
                        table.Dispose();
                    _data.Dispose();
                }
            }
        }
        /// <summary>
		/// Releases all unmanaged resources.
		/// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        #endregion
    }
}
